ext/ldap/ldap.c

```diff
@@ -2491,351 +2491,255 @@ PHP_FUNCTION(ldap_delete_ext)
-		}
-	}
-
-	return max_len;
-}
-/* }}} */
-
-/* {{{ _ldap_hash_fetch */
-static void _ldap_hash_fetch(zval *hashTbl, const char *key, zval **out)
-{
-	*out = zend_hash_str_find(Z_ARRVAL_P(hashTbl), key, strlen(key));
-}
-/* }}} */
-
 /* {{{ Perform multiple modifications as part of one operation */
 PHP_FUNCTION(ldap_modify_batch)
 {
-	zval *serverctrls = NULL;
+	zval *server_controls_zv = NULL;
 	ldap_linkdata *ld;
-	zval *link, *mods, *mod, *modinfo;
-	zend_string *modval;
-	zval *attrib, *modtype, *vals;
-	zval *fetched;
+	zval *link;
 	char *dn;
 	size_t dn_len;
-	int i, j, k;
-	int num_mods, num_modprops, num_modvals;
+	HashTable *modifications;
 	LDAPMod **ldap_mods;
 	LDAPControl **lserverctrls = NULL;
-	uint32_t oper;
 
 	/*
-	$mods = array(
-		array(
+	$mods = [
+		[
 			"attrib" => "unicodePwd",
 			"modtype" => LDAP_MODIFY_BATCH_REMOVE,
-			"values" => array($oldpw)
-		),
-		array(
+			"values" => [$old_pwd]
+		],
+		[
 			"attrib" => "unicodePwd",
 			"modtype" => LDAP_MODIFY_BATCH_ADD,
-			"values" => array($newpw)
-		),
-		array(
+			"values" => [$new_pwd]
+		],
+		[
 			"attrib" => "userPrincipalName",
 			"modtype" => LDAP_MODIFY_BATCH_REPLACE,
-			"values" => array("janitor@corp.contoso.com")
-		),
-		array(
+			"values" => ["janitor@corp.contoso.com"]
+		],
+		[
 			"attrib" => "userCert",
 			"modtype" => LDAP_MODIFY_BATCH_REMOVE_ALL
-		)
-	);
+		],
+	];
 	*/
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "Osa/|a!", &link, ldap_link_ce, &dn, &dn_len, &mods, &serverctrls) != SUCCESS) {
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "Osh/|a!", &link, ldap_link_ce, &dn, &dn_len, &modifications, &server_controls_zv) != SUCCESS) {
 		RETURN_THROWS();
 	}
 
 	ld = Z_LDAP_LINK_P(link);
 	VERIFY_LDAP_LINK_CONNECTED(ld);
 
 	/* perform validation */
-	{
-		zend_string *modkey;
-		zend_long modtype;
+	/* make sure the DN contains no NUL bytes */
+	if (zend_char_has_nul_byte(dn, dn_len)) {
+		zend_argument_value_error(2, "must not contain null bytes");
+		RETURN_THROWS();
+	}
 
-		/* to store the wrongly-typed keys */
-		zend_ulong tmpUlong;
+	if (zend_hash_num_elements(modifications) == 0) {
+		zend_argument_must_not_be_empty_error(3);
+		RETURN_THROWS();
+	}
+	if (!zend_array_is_list(modifications)) {
+		zend_argument_value_error(3, "must be a list");
+		RETURN_THROWS();
+	}
 
-		/* make sure the DN contains no NUL bytes */
-		if (_ldap_strlen_max(dn, dn_len) != dn_len) {
-			zend_argument_type_error(2, "must not contain null bytes");
+	zval *modification_zv = NULL;
+	ZEND_HASH_FOREACH_VAL(modifications, modification_zv) {
+		if (Z_TYPE_P(modification_zv) != IS_ARRAY) {
+			zend_argument_type_error(3, "must only contain arrays");
 			RETURN_THROWS();
 		}
 
-		/* make sure the top level is a normal array */
-		zend_hash_internal_pointer_reset(Z_ARRVAL_P(mods));
-		if (zend_hash_get_current_key_type(Z_ARRVAL_P(mods)) != HASH_KEY_IS_LONG) {
-			zend_argument_type_error(3, "must be integer-indexed");
+		SEPARATE_ARRAY(modification_zv);
```

Comment:
I am not sure why we would need to separate the array as we are just doing validation and reading from it.

Reply:
Because later on the code uses `convert_to_*` calls, but that means the zval may get transformed causing either corruption with opcache; or spooky unexpected changes at a distance.